backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropLowering.kt

```diff
@@ -260,9 +265,9 @@ internal class InteropLoweringPart1(val context: Context) : BaseInteropIrTransfo
             IrDeclarationOriginImpl("OVERRIDING_INITIALIZER_BY_CONSTRUCTOR")
 
     private fun IrConstructor.overridesConstructor(other: IrConstructor): Boolean {
-        return this.descriptor.valueParameters.size == other.descriptor.valueParameters.size &&
-                this.descriptor.valueParameters.all {
-                    val otherParameter = other.descriptor.valueParameters[it.index]
+        return this.valueParameters.size == other.valueParameters.size &&
```

Comment:
Does this particular patch fixes anything?
This predicate is technically related to frontend, so I'm not sure it is worth rewriting in terms of IR.

Reply:
This patch fixes using the `-Xinclude` option for a library containing a class constructor annotated with `@OverrideInit`. E.g. if the following class is compiled into a klib:
```kotlin
@ExportObjCClass
class ViewController : NSObject {
    @OverrideInit constructor() : super()
}
```
and then passed to the compiler using the `-Xinclude` option, [this](https://github.com/JetBrains/kotlin-native/pull/3295/files/7a70e1e0833aed3c7282108335a5976f2b1e8906#diff-4fbff1fd1ac95615545d03980ba283a4R1008) check based on descriptors fails:
```
error: receiver of 'initBy' must be a 'this' of the constructed class
```

Migrating to IR fixes this issue. I agree that such checks should be done by the frontend. But currently switching to IR here is the simplest way to fix this problem. It's important because I'm going to enable the two-stage compilation of final binaries in the MPP plugin by default.

Actually to fix this issue only aforementioned check should be switched to IR. But migrating to IR other parts of this lowering is usually simple (and I didn't change parts where such migrating is more or less complex) and may help us to avoid similar failures on other checks.

Reply:
Opened a ticket to save the reproducing steps: https://youtrack.jetbrains.com/issue/KT-33645.

Reply:
I'm not asking about the entire PR, but only about this particular change in `overridesConstructor`.
What does it fix?

> and may help us to avoid similar failures on other checks.

But will complicate moving this checks to frontend.
I somewhat dislike fixing the code that is not broken.

Also note that the actual bug is likely located in wrapped descriptors implementation.

Reply:
> What does it fix?

Nothing particular, just a refactoring.

> But will complicate moving this checks to frontend.

Sounds reasonable. Ok, I'll return descriptors in all places you marked as belonging to the frontend.

> Also note that the actual bug is likely located in wrapped descriptors implementation.

As far as I understand the preferred approach in such cases is to migrate from wrapped descriptors to IR instead of fixing wrapped descriptors.